jax/_src/scipy/optimize/lbfgs.py

```diff
@@ -0,0 +1,229 @@
+    s_history=jnp.zeros((maxcor, d), dtype=dtype),
+    y_history=jnp.zeros((maxcor, d), dtype=dtype),
+    rho_history=jnp.zeros((maxcor,), dtype=dtype),
+    gamma=1.,
+    status=0,
+    ls_status=0,
+  )
+
+  def cond_fun(state: LBFGSResults):
+    return (~state.converged) & (~state.failed)
+
+  def body_fun(state: LBFGSResults):
+    # find search direction
+    p_k = _two_loop_recursion(state)
+
+    # line search
+    ls_results = line_search(
+      f=fun,
+      xk=state.x_k,
+      pk=p_k,
+      old_fval=state.f_k,
+      gfk=state.g_k,
+      maxiter=maxls,
+    )
+
+    # evaluate at next iterate
+    s_k = ls_results.a_k * p_k
+    x_kp1 = state.x_k + s_k
+    f_kp1 = ls_results.f_k
+    g_kp1 = ls_results.g_k
+    y_k = g_kp1 - state.g_k
+    rho_k_inv = jnp.real(_dot(y_k, s_k))
+    rho_k = jnp.reciprocal(rho_k_inv)
+    gamma = rho_k_inv / jnp.real(_dot(jnp.conj(y_k), y_k))
+
+    # replacements for next iteration
+    status = 0
+    status = jnp.where(state.f_k - f_kp1 < ftol, 4, status)
+    status = jnp.where(state.ngev >= maxgrad, 3, status)
+    status = jnp.where(state.nfev >= maxfun, 2, status)
+    status = jnp.where(state.k >= maxiter, 1, status)
+    status = jnp.where(ls_results.failed, 5, status)
+
+    converged = jnp.linalg.norm(g_kp1, ord=norm) < gtol
+
+    state = state._replace(
+      converged=converged,
+      failed=(status > 0) & (~converged),
+      k=state.k + 1,
+      nfev=state.nfev + ls_results.nfev,
+      ngev=state.ngev + ls_results.ngev,
+      x_k=x_kp1,
+      f_k=f_kp1,
+      g_k=g_kp1,
+      s_history=_update_history_vectors(history=state.s_history, new=s_k),
+      y_history=_update_history_vectors(history=state.y_history, new=y_k),
+      rho_history=_update_history_scalars(history=state.rho_history, new=rho_k),
+      gamma=gamma,
+      status=jnp.where(converged, 0, status),
+      ls_status=ls_results.status,
+    )
+
+    return state
+
+  return lax.while_loop(cond_fun, body_fun, state_initial)
+
+
+def _two_loop_recursion(state: LBFGSResults):
+  his_size = len(state.rho_history)
+  curr_size = jnp.where(state.k < his_size, state.k, his_size)
+  q = -jnp.conj(state.g_k)
+  a_his = jnp.zeros_like(state.rho_history)
+
+  def body_fun1(j, carry):
+    i = his_size - 1 - j
+    _q, _a_his = carry
+    a_i = state.rho_history[i] * jnp.real(_dot(jnp.conj(state.s_history[i]), _q))
+    _a_his = ops.index_update(_a_his, ops.index[i], a_i)
+    _q = _q - a_i * jnp.conj(state.y_history[i])
+    return _q, _a_his
+
+  q, a_his = lax.fori_loop(0, curr_size, body_fun1, (q, a_his))
+  q = state.gamma * q
+
+  def body_fun2(j, _q):
+    i = his_size - curr_size + j
+    b_i = state.rho_history[i] * jnp.real(_dot(state.y_history[i], _q))
+    _q = _q + (a_his[i] - b_i) * state.s_history[i]
+    return _q
+
+  q = lax.fori_loop(0, curr_size, body_fun2, q)
+  return q
+
+
+def _update_history_vectors(history, new):
+  return ops.index_update(jnp.roll(history, -1, axis=0), ops.index[-1, :], new)
```

Comment:
Don't bother to update this now, but I suspect it would be slightly more efficient to use a rolling buffer rather to roll the entire `history` matrix in each update.

Reply:
I do not know how to do this.
Adding a TODO comment for now.